File: 10_short_rate_models_with_stochastic_volatility/python/stochastic_vol_model.py
"""
stochastic_vol_model.py

Chapter 10 — Short Rate Models with Stochastic Volatility.

This model extends the Chapter 09 Hull-White model:

    dr_t = [theta(t) - a r_t] dt + sigma dW_t

to a stochastic-volatility version:

    dr_t = [theta(t) - a r_t] dt + sqrt(v_t) dW_t^r

    dv_t = kappa (v_bar - v_t) dt + eta sqrt(v_t) dW_t^v

with:

    corr(dW_t^r, dW_t^v) = rho dt

The model reuses the Chapter 09 HullWhiteModel object for:
- theta(t)
- mean reversion a
- initial curve
"""

import numpy as np


class StochasticVolShortRateModel:
    """
    Hull-White-style short-rate model with stochastic variance.

    Parameters
    ----------
    base_hw_model : HullWhiteModel
        Chapter 09 Hull-White model. We reuse theta(t), curve and a.

    kappa : float
        Variance mean reversion speed.

    v_bar : float
        Long-run variance level.

    eta : float
        Volatility of variance.

    rho : float
        Correlation between short-rate shock and variance shock.
    """

    def __init__(
        self,
        base_hw_model,
        kappa,
        v_bar,
        eta,
        rho
    ):
        self.base_hw_model = base_hw_model
        self.a = base_hw_model.a
        self.curve = base_hw_model.curve

        self.kappa = kappa
        self.v_bar = v_bar
        self.eta = eta
        self.rho = rho

        if self.kappa <= 0:
            raise ValueError("kappa must be positive.")

        if self.v_bar <= 0:
            raise ValueError("v_bar must be positive.")

        if self.eta < 0:
            raise ValueError("eta must be non-negative.")

        if self.rho <= -1.0 or self.rho >= 1.0:
            raise ValueError("rho must be between -1 and 1.")

    def theta(self, t):
        """
        Reuse Hull-White theta(t) from Chapter 09.
        """

        return self.base_hw_model.theta(t)

    def short_rate_drift(self, t, r):
        """
        Short-rate drift:

            theta(t) - a r_t
        """

        return self.theta(t) - self.a * r

    def variance_drift(self, v):
        """
        CIR-style variance drift:

            kappa (v_bar - v_t)
        """

        return self.kappa * (self.v_bar - v)

    def evolve(
        self,
        t,
        r,
        v,
        dt,
        z_rate,
        z_vol
    ):
        """
        Evolve short rate and variance by one Euler step.

        Full truncation is used to keep variance non-negative.
        """

        v_pos = max(v, 0.0)

        r_next = (
            r
            +
            self.short_rate_drift(t, r) * dt
            +
            np.sqrt(v_pos) * np.sqrt(dt) * z_rate
        )

        v_next = (
            v
            +
            self.variance_drift(v_pos) * dt
            +
            self.eta * np.sqrt(v_pos) * np.sqrt(dt) * z_vol
        )

        v_next = max(v_next, 0.0)

        return r_next, v_next

    def instantaneous_volatility(self, v):
        """
        Convert variance into instantaneous short-rate volatility.
        """

        return np.sqrt(max(v, 0.0))

File: 10_short_rate_models_with_stochastic_volatility/cpp/StochasticVolModel.cpp
#include "StochasticVolModel.hpp"

#include <cmath>
#include <stdexcept>
#include <algorithm>

/*
    Constructor
*/

StochasticVolModel::StochasticVolModel(
    HullWhiteModel* baseModel_,
    double kappa_,
    double vBar_,
    double eta_,
    double rho_
){
    baseModel = baseModel_;
    a = baseModel_->a;

    kappa = kappa_;
    vBar = vBar_;
    eta = eta_;
    rho = rho_;

    if(kappa <= 0.0){
        throw std::runtime_error("kappa must be positive.");
    }

    if(vBar <= 0.0){
        throw std::runtime_error("vBar must be positive.");
    }

    if(eta < 0.0){
        throw std::runtime_error("eta must be non-negative.");
    }

    if(rho <= -1.0 || rho >= 1.0){
        throw std::runtime_error("rho must be between -1 and 1.");
    }
}

/*
    Reuse theta(t) from Chapter 09 Hull-White model.
*/

double StochasticVolModel::theta(
    double t
) const {

    return baseModel->theta(t);
}

/*
    Short-rate drift:

        theta(t) - a r_t
*/

double StochasticVolModel::shortRateDrift(
    double t,
    double r
) const {

    return
        theta(t)
        -
        a * r;
}

/*
    Variance drift:

        kappa (vBar - v_t)
*/

double StochasticVolModel::varianceDrift(
    double v
) const {

    return
        kappa
        *
        (vBar - v);
}

/*
    Instantaneous volatility:

        sqrt(max(v,0))
*/

double StochasticVolModel::instantaneousVolatility(
    double v
) const {

    return
        std::sqrt(
            std::max(
                v,
                0.0
            )
        );
}

/*
    One Euler step with full truncation.

    v_pos = max(v,0)

    r_{t+dt}
        =
        r_t
        +
        [theta(t) - a r_t] dt
        +
        sqrt(v_pos) sqrt(dt) Z_r

    v_{t+dt}
        =
        v_t
        +
        kappa(vBar - v_pos) dt
        +
        eta sqrt(v_pos) sqrt(dt) Z_v
*/

void StochasticVolModel::evolve(
    double t,
    double r,
    double v,
    double dt,
    double zRate,
    double zVol,
    double& rNext,
    double& vNext
) const {

    double vPos =
        std::max(
            v,
            0.0
        );

    rNext =
        r
        +
        shortRateDrift(t, r) * dt
        +
        std::sqrt(vPos)
        *
        std::sqrt(dt)
        *
        zRate;

    vNext =
        v
        +
        varianceDrift(vPos) * dt
        +
        eta
        *
        std::sqrt(vPos)
        *
        std::sqrt(dt)
        *
        zVol;

    vNext =
        std::max(
            vNext,
            0.0
        );
}
